diff --git a/app/serializers/manifest_serializer.rb b/app/serializers/manifest_serializer.rb
index 48f3aa7a6..4a0887feb 100644
--- a/app/serializers/manifest_serializer.rb
+++ b/app/serializers/manifest_serializer.rb
@@ -5,14 +5,7 @@ class ManifestSerializer < ActiveModel::Serializer
   include ActionView::Helpers::TextHelper

   ICON_SIZES = %w(
-    36
-    48
-    72
-    96
-    144
     192
-    256
-    384
     512
   ).freeze
